File: Python/ByTier/A/1868_A_Fill_in_the_Matrix.py
# Problem: CF 1868 A - Fill in the Matrix
# https://codeforces.com/contest/1868/problem/A

"""
Code Purpose:
This code solves the problem of filling an n x m matrix with permutations of length m such that the beauty of the matrix (defined as the MEX of column values) is maximized.

Algorithms/Techniques:
- Greedy approach with pattern construction
- Utilizes the property that MEX of a set of numbers is the smallest non-negative integer not present in the set
- Constructs rows to ensure that column MEX values are as varied as possible

Time Complexity: O(n * m) - We iterate through all rows and columns once to construct the matrix.
Space Complexity: O(n * m) - The space required to store the matrix and output.
"""

import sys

for _ in range(int(input())):
    l = sys.stdin.readline().split()
    n = int(l[0])
    m = int(l[1])
    if m == 1:
        print(0)
        for i in range(n):
            sys.stdout.write(str(0) + "\n")
        continue
    k = min(m, n + 1)  # Determine the size of the repeating pattern
    sys.stdout.write(str(k) + "\n")
    # Generate first k-1 rows using a cyclic shift pattern to maximize diversity
    for i in range(k - 1):
        sys.stdout.write(
            " ".join(
                [str((i + j) % k) for j in range(k)] + [str(j) for j in range(k, m)]
            )
            + "\n"
        )
    # Fill remaining rows with a simple permutation 0..m-1
    for i in range(k - 1, n):
        sys.stdout.write(" ".join([str(j) for j in range(m)]) + "\n")


# https://github.com/VaHiX/CodeForces/

File: CPP/ByRound/1868/1868_A_Fill_in_the_Matrix.cpp
// Problem: CF 1868 A - Fill in the Matrix
// https://codeforces.com/contest/1868/problem/A

/*
Code Purpose:
This program fills an n x m matrix such that the beauty of the matrix (defined as the MEX of column values) is maximized.
Each row is a permutation of integers from 0 to m-1.
The algorithm uses a cyclic shifting technique to construct rows that distribute values evenly across columns,
ensuring high diversity in column MEX values.

Algorithms/Techniques:
- Cyclic shifting of column indices to create permutations
- Mathematically designed row construction to maximize MEX across columns

Time Complexity: O(n * m) - Each element of the matrix is computed once.
Space Complexity: O(1) - Only a constant amount of extra space is used (excluding output storage).
*/

#include <stdio.h>
#include <algorithm>

using namespace std;
int n, m, i, j, T_T;
int main() {
  for (scanf("%d", &T_T); T_T--;) {
    scanf("%d%d", &n, &m);
    // Output the maximum possible beauty of the matrix.
    // If m > 1, maximum beauty is min(n + 1, m), otherwise it's 0.
    printf("%d\n", m > 1 ? min(n + 1, m) : 0);
    // Generate each row of the matrix.
    for (i = 1; i <= n; ++i, puts(""))
      // For each row, compute the value at column j using cyclic shift.
      for (j = 1; j <= m; ++j)
        printf("%d ", (min(i, m - 1) + j) % m);
  }
}


// https://github.com/VaHiX/CodeForces/